processor/src/main/java/org/mapstruct/ap/internal/services/Services.java

```diff
@@ -29,53 +28,18 @@
  */
 public class Services {
 
-    private static final ConcurrentMap<Class<?>, Object> SERVICES = new ConcurrentHashMap<Class<?>, Object>();
-
     private Services() {
     }
 
     public static <T> T get(Class<T> serviceType, T defaultValue) {
-        @SuppressWarnings("unchecked")
-        T service = (T) SERVICES.get( serviceType );
-
-        if ( service == null ) {
-            service = loadAndCache( serviceType, defaultValue );
-        }
 
-        return service;
-    }
+        Iterator<T> services = ServiceLoader.load( serviceType, Services.class.getClassLoader() ).iterator();
 
-    private static <T> T loadAndCache(Class<T> serviceType, T defaultValue) {
-        T service = find( serviceType );
-        if ( service == null ) {
-            service = defaultValue;
+        if ( services.hasNext() ) {
+            return services.next();
         }
-
-        @SuppressWarnings("unchecked")
-        T cached = (T) SERVICES.putIfAbsent( serviceType, service );
-        if ( cached != null ) {
-            service = (T) cached;
+        else {
+            return defaultValue;
         }
-
-        return service;
-    }
-
-    private static <T> T find(Class<T> spi) {
-        T matchingImplementation = null;
-
-        for ( T implementation : ServiceLoader.load( spi, spi.getClassLoader() ) ) {
-            if ( matchingImplementation == null ) {
-                matchingImplementation = implementation;
-            }
-            else {
-                throw new IllegalStateException(
-                    "Multiple implementations have been found for the service provider interface "
```

Comment:
Hum, seems we won't raise an exception anymore if there are multiple service implementations? That check was quite meaningful IMO.